rfcs/0015-const-generics.md

```diff
@@ -0,0 +1,299 @@
+
+A simple example would be:
+
+```rust
+fn id<const SIZE: usize>(array: [u64; SIZE]) -> [u64; SIZE] {
+    array
+}
+```
+
+This also allows `impl` items such as
+
+```rust
+impl<const N: usize> AbiEncode for str[N] {
+    ...
+}
+```
+
+This constant can be inferred or explicitly specified. When inferred, the syntax is no different than just using the item:
+
+```rust
+id([1u8])
+```
+
+In the example above, the Sway compiler will infer `SIZE` to be one, because `id` parameter will be infered to be `[u8; 1]`.
+
+For the cases where the compiler cannot infer this value, or this value comes from an expression, it is possible to do:
+
+```rust
+id::<1>([1u8]);
+id::<{1 + 1}>([1u8, 2u8]);
+```
+
+When the value is not a literal, but an expression, it is named "const generic expression" and it needs to be enclosed by curly braces. This will fail if the expression cannot be evaluated as `const`.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+This new syntax has three forms: declarations, instantiations and references.
+
+"const generics declarations" can appear anywhere all other generic arguments declarations are valid:
+
+1. Function/method declaration;
+1. Struct/Enum declaration;
+1. `impl` declarations.
+
+```rust
+// 1
+fn id<const N: usize>(...) { ... }
+
+// 2
+struct SpecialArray<const N: usize> {
+    inner: [u8; N],
+}
+
+//3
+impl<const N: usize> AbiEncode for str[N] {
+    ...
+}
+```
+
+"const generics instantiations" can appear anywhere all other generic argument instantiations are valid:
+
+1. Function/method reference;
+1. Struct/Enum reference;
+1. Fully qualified call paths.
+
+```rust
+// 1
+id::<1>([1u8]);
+special_array.do_something::<1>();
+
+// 2
+SpecialArray::<1>::new();
+
+// 3
+<SpecialArray::<1> as SpecialArrayTrait::<1>>::f();
+```
+
+"const generics references" can appear anywhere any other identifier appears. For semantic purposes, there is no difference between the reference of a "const generics" and a "normal" const.
+
+```rust
+fn f<const I: usize>() {
+    __log(I);
+}
+```
+
+Different from type generics, const generics cannot appear at:
+
+1. constraints;
+1. where types are expected.
+
+```rust
+// 1 - INVALID
+fn f<const I: usize>() where I > 10 {
+    ...
+}
+
+// 2 - INVALID
+fn f<const I: usize>() -> Vec<I> {
+    Vec::<I>::new()
+}
+```
+
+## Const Value Specialization
+
+By the nature of `impl` declarations, it is possible to specialize for some specific types. For example:
+
+```rust
+impl SomeStruct<bool> {
+    fn f() {        
+ }
+}
+```
+
+In the example above, `f` only is available when the generic argument of `SomeStruct` is known to be `bool`. This will not be supported for "const generics", which means that the example below will not be supported:
+
+```rust
+impl SomeIntegerStruct<1> {
+    ...
+}
+
+impl SomeBoolStruct<false> {
```

Comment:
As mentioned, we should explicitly specify which types are supported. Out of this example with `bool` I assume all uint types + `bool` will be supported. I don't see a reason not to.

Reply:
The idea of this part, is to inform that specialization will not be supported for "const generics". The "type specialization" is just an example, and it already exists in Sway.